flutter-engine/src/texture_registry.rs

```diff
@@ -1,96 +1,52 @@
-};
-
-use gl::types::*;
-use log::debug;
-
 use crate::ffi::ExternalTexture as FlutterTexture;
-use crate::ffi::ExternalTextureFrame;
+use crate::ffi::ExternalTextureFrame as FlutterFrame;
 use crate::FlutterEngine;
+use image::RgbaImage;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::sync::atomic::AtomicI64;
+use std::sync::{Arc, Mutex};
 
 type TextureID = i64;
 
 #[derive(Default)]
 pub struct TextureRegistry {
-    textures: TextureStore,
-}
-
-#[derive(Default)]
-struct TextureStore {
-    /// Stores freshly registered textures without an OpenGL texture attached.
-    initial: HashMap<TextureID, Arc<ExternalTexture>>,
-    /// Stores created textures that may be dropped in other locations.
-    created: HashMap<TextureID, Weak<ExternalTexture>>,
+    textures: Arc<Mutex<HashMap<TextureID, ExternalTexture>>>,
 }
 
 pub struct ExternalTexture {
     texture: FlutterTexture,
-    texture_data: Mutex<Option<TextureData>>,
+    frame: FlutterFrame,
 }
 
-struct TextureData {
-    name: GLuint,
-    width: u32,
-    height: u32,
-}
+unsafe impl Send for ExternalTexture {}
 
 impl TextureRegistry {
     pub fn new() -> Self {
         Default::default()
     }
 
-    pub fn create_texture(&mut self, engine: &FlutterEngine) -> Arc<ExternalTexture> {
+    pub fn create_texture(&self, engine: &FlutterEngine, img: RgbaImage) -> TextureID {
         static TEXTURE_ID: AtomicI64 = AtomicI64::new(1);
 
         let texture_id = TEXTURE_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
-        let flutter_texture = engine.register_external_texture(texture_id);
-
-        let texture = Arc::new(ExternalTexture {
-            texture: flutter_texture,
-            texture_data: Mutex::new(None),
-        });
+        let textures = self.textures.clone();
+        engine.run_on_platform_thread(move |engine| {
+            let flutter_texture = engine.register_external_texture(texture_id);
 
-        self.textures
-            .initial
-            .insert(texture_id, Arc::clone(&texture));
+            engine.run_on_render_thread(move |_engine| {
+                let flutter_frame = unsafe { create_gl_texture(texture_id, img) };
 
-        texture
-    }
+                let texture = ExternalTexture {
+                    texture: flutter_texture,
+                    frame: flutter_frame,
+                };
 
-    pub fn get_texture_frame(
-        &mut self,
-        texture_id: TextureID,
-        (width, height): (u32, u32),
-    ) -> Option<ExternalTextureFrame> {
-        if let Some(texture) = self.textures.initial.remove(&texture_id) {
-            // texture is still initial --> create it
-            let frame = unsafe { create_gl_texture(texture_id, (width, height)) };
-            let mut data = texture.texture_data.lock().unwrap();
-            data.replace(TextureData {
-                name: frame.name,
-                width,
-                height,
+                texture.mark_frame_available();
```

Comment:
We should insert the texture into the map before making the frame available, otherwise there is a possible race condition.